### PR TITLE
Stabilize math in SpanningTree.log_partition_function

### PR DESCRIPTION
**Pull request by fritzo**
This makes the numerical stabilization math more fine-grained in `SpanningTree.log_partition_function`.

Previously we shifted by a max over all edges. After this PR we shift by column sums. This is advantageous because the resulting Laplacian matrix has constant 1's on the diagonal and is safer for the Cholesky factorization in `.logdet()`.

Could a reviewer please go over the math in detail to ensure gradients are correct? Thanks in advance!

## Tested
- [x] refactoring covered by existing tests
- [x] fixes nan issues in a gradient-based inference that uses `.log_partition_function`

### CODE REVIEW

File: pyro/distributions/spanning_tree.cpp [hunk 1]
```diff
@@ -142,7 +142,7 @@ at::Tensor sample_tree_approx(at::Tensor edge_logits) {
   // the complete graph. The id of an edge (v1,v2) is k = v1+v2*(v2-1)/2.
   auto edge_ids = torch::empty({E}, at::kLong);
   // This maps each vertex to whether it is a member of the cumulative tree.
-  auto components = torch::zeros({V}, at::kByte);
+  auto components = torch::zeros({V}, at::kBool);
```
Comment (fritzo) [inline pyro/distributions/spanning_tree.cpp:145]: this avoids a deprecation warning in PyTorch 1.7

File: pyro/distributions/spanning_tree.py [hunk 2]
```diff
@@ -112,20 +112,21 @@ def log_partition_function(self):
         # use a Cholesky decomposition to compute the log determinant.
         # See https://en.wikipedia.org/wiki/Kirchhoff%27s_theorem
         V = self.num_vertices
-        grid = make_complete_graph(V)
-        shift = self.edge_logits.detach().max()
-        edge_probs = (self.edge_logits - shift).exp()
```
Comment (martinjankowiak) [inline pyro/distributions/spanning_tree.py:117]: these are normalized along each column/row?

File: pyro/distributions/spanning_tree.py [hunk 3]
```diff
@@ -112,20 +112,21 @@ def log_partition_function(self):
...
-        adjacency[grid[1], grid[0]] = edge_probs
-        laplacian = adjacency.sum(-1).diag() - adjacency
+        v1, v2 = make_complete_graph(V).unbind(0)
+        logits = self.edge_logits.new_full((V, V), -math.inf)
+        logits[v1, v2] = self.edge_logits
+        logits[v2, v1] = self.edge_logits
+        log_diag = logits.logsumexp(-1)
+        # Numerically stabilize so that laplacian has 1's on the diagonal.
+        shift = 0.5 * log_diag
+        laplacian = torch.eye(V) - (logits - shift - shift[:, None]).exp()
```
Comment (martinjankowiak) [inline pyro/distributions/spanning_tree.py:122]: looks right to me. what's the deal with the truncation?

Comment (fritzo) [reply pyro/distributions/spanning_tree.py:122]: That's just how the theorem goes: you delete any single column and row, then compute the determinant.
See https://en.wikipedia.org/wiki/Kirchhoff%27s_theorem#An_example_using_the_matrix-tree_theorem

File: pyro/distributions/spanning_tree.py (hunk 2 above)
Comment (fritzo) [reply pyro/distributions/spanning_tree.py:117]: That's right, I'm normalizing by column, but since the matrix is symmetric, I split that normalization to be half along rows and half along columns. The diagonal doesn't care whether it's all-column or all-row or half-and-half.

**Review by martinjankowiak — APPROVED** [review]
